### PR TITLE
Refactor CI builds and add unit testing for IronPython

### PR DESCRIPTION
**Pull request by gonzalocasas**
Since Python 2.7 support is being dropped by all our dependencies, we cannot keep building on it but since we need to keep supporting IronPython 2.7 for the foreseeable future, we must ensure that at least syntax is 2.7 compatible so that it runs in Rhino and friends.

In this PR:
 - Added a reworked travis configuration to have:
   - One build on each of the 3 main OSes (Linux, Mac, Windows) on the latest Python stable release available. (currently, 3.8 on win/linux, and 3.7 on mac)
    - One build on python 3.6 (linux) because it is still probably the most widely used version.
    - One build on IronPython 2.7.8 (windows). This is the same as the version embedded on Rhino 6 for Windows (not on Mac, that uses 2.7.9 and some differences affect the behavior of buffers, in particular, visible on the glTF support)
  - Added unit test runner for IronPython. To this purpose, a tiny `pytest` clone was added, because `pytest` does not support IronPython. It's a minimal tool but it is able to run all tests we have creating an output that resembles that of `pytest`.
  - Fixed some bugs uncovered by running the tests on IronPython:
    - Added precision to the `Mesh.from_stl` to fix mesh test that was collapsing some vertices on ipy
    - Added fallback to the incomplete bytearray support of IronPython 2.7.8 for the glTF exporter
    - Tweaked the `compas.geometry.allclose` function to also compare recursively, just like the `numpy.allclose` so that lots of tests can use that instead of numpy's version, and hence run on IronPython.
  - Skipped two tests that depend on `planarity`

### What type of change is this?

- [x] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.

### Checklist

1. [ ] ~~Add the change to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading: `Added`, `Changed`, `Removed`.~~
1. [x] Run all tests on your computer (i.e. `invoke test`).
1. [ ] If you add new functions/classes, check that:
   1. [ ] Are available on a second-level import, e.g. `compas.datastructures.Mesh`.
   1. [ ] Add unit tests (especially important for algorithm implementations).

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -5,7 +5,8 @@ ezdxf
 imageio <= 2.6; python_version < '3.5'
 imageio >= 2.7; python_version >= '3.5'
 matplotlib >= 2.2, < 3.0; python_version >= '2.7' and python_version < '3.0'
-matplotlib >= 2.2, < 3.1; python_version >= '3.5' and sys_platform == 'win32'
+matplotlib >= 2.2, < 3.1; python_version >= '3.5' and python_version <= '3.7' and sys_platform == 'win32'
+matplotlib >= 3.1; python_version >= '3.8' and sys_platform == 'win32'
```
Comment (brgcode) [inline requirements.txt:9]: nice :)

File: requirements.txt [hunk 2]
```diff
@@ -5,7 +5,8 @@ ezdxf
...
 matplotlib >= 2.2; python_version >= '3.5' and sys_platform != 'win32'
```
Comment (tomvanmele) [inline requirements.txt:10]: :)

File: src/compas/geometry/_core/basic.py [hunk 3]
```diff
@@ -110,7 +110,8 @@ def allclose(l1, l2, tol=1e-05):
     .. [1] https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html
 
     """
-    if any(fabs(a - b) > tol for a, b in zip(l1, l2)):
+
+    if any(not allclose(a, b, tol) if hasattr(a, '__iter__') else fabs(a - b) > tol for a, b in zip(l1, l2)):
```
Comment (tomvanmele) [inline src/compas/geometry/_core/basic.py:114]: ?

Comment (tomvanmele) [reply src/compas/geometry/_core/basic.py:114]: which extreme boundary case are we accommodating here?

File: tests/compas/datastructures/test_mesh.py [hunk 4]
```diff
@@ -49,12 +49,12 @@ def test_from_ply():
 
 
 def test_from_stl():
-    mesh = Mesh.from_stl(compas.get('cube_ascii.stl'))
+    mesh = Mesh.from_stl(compas.get('cube_ascii.stl'), '6f')
```
Comment (tomvanmele) [inline tests/compas/datastructures/test_mesh.py:52]: still a question for me how do reliably deal with precision. this is btw, why i started adding header info to the files that we write... i think precision should be part of that...

File: tests/compas/geometry/test_basics.py [hunk 5]
```diff
@@ -34,7 +33,7 @@
 ]
 )
 def test_angle_vectors(u, v, angle):
-    assert angle_vectors(u, v) == pytest.approx(angle)
+    assert close(angle_vectors(u, v), angle)
```
Comment (tomvanmele) [inline tests/compas/geometry/test_basics.py:36]: so you prefer that we rely on our own interpretation of "close enough"?

File: tests/compas/geometry/test_bbox.py [hunk 6]
```diff
@@ -50,5 +54,11 @@ def test_oriented_bounding_box_xy_numpy(coords, expected):
...
-    print(oriented_bounding_box_numpy(coords).tolist())
-    assert expected == oriented_bounding_box_numpy(coords).tolist()
+    if compas.IPY:
+        return
+
+    from compas.geometry import oriented_bounding_box_numpy
+
+    results = oriented_bounding_box_numpy(coords).tolist()
+    for result, expected_values in zip(results, expected):
+        assert allclose(result, expected_values, tol=1e-3)
```
Comment (tomvanmele) [inline tests/compas/geometry/test_bbox.py:64]: we should have a general strategy for precision...

**Review by tomvanmele — COMMENTED** [review]
precision seems to be a thing

File: src/compas/geometry/_core/basic.py (hunk 3 above)
Comment (gonzalocasas) [reply src/compas/geometry/_core/basic.py:114]: is not a boundary case at all, this recursively compares if the element is an iterable, ie `list of list of floats` can be compared. This is how `np.allclose` behaves. The reasons for replicating the behavior is not just consistency, but actually the need to remove all `numpy` from a lot of unit tests that were using `np.allclose` to check results (but were not using `numpy` at all other than that).

File: tests/compas/datastructures/test_mesh.py (hunk 4 above)
Comment (gonzalocasas) [reply tests/compas/datastructures/test_mesh.py:52]: Headers sounds like a sound idea ;)  Additionally, the serialization formats used by compas should have a schema somewhere, and support versioning of themselves; but all that is a discussion for another PR.

File: tests/compas/geometry/test_basics.py (hunk 5 above)
Comment (gonzalocasas) [reply tests/compas/geometry/test_basics.py:36]: Not really, again and just like the change of `np.allclose` to `compas.geometry.allclose`, the idea is was only to make the tests run. In this case, using `pytest.approx` was not really adding a lot of benefit than what we could easily do with our `close`, and implementing `approx` in my pytest clone would have been too much work (it seems to rely on a lot of magic methods to do its magic).

File: src/compas/geometry/_core/basic.py (hunk 3 above)
Comment (tomvanmele) [reply src/compas/geometry/_core/basic.py:114]: coolio
